text/3691-trusted-publishing-cratesio.md

```diff
@@ -0,0 +1,384 @@
+
+These same API tokens are likely stored long term in a secrets manager like GitHub Actions Secrets or AWS Secrets Manager where a misconfiguration could lead to exposure of the secret.
+
+> [!NOTE] 
+> Other security challenges that would persist by not utilizing Trusted Publishing are detailed in the [Motivation](#motivation) section of this document.
+
+## Prior art
+[prior-art]: #prior-art
+
+Other package repositories have implemented similar support:
+
+- [PyPI's Trusted Publishing](https://docs.pypi.org/trusted-publishers/)
+- [RubyGem's Trusted Publishing](https://guides.rubygems.org/trusted-publishing/)
+- [Dart's Automated Publishing](https://dart.dev/tools/pub/automated-publishing)
+- [OSSF's Trusted Publishers For All Package Repositories Blog Post](https://repos.openssf.org/trusted-publishers-for-all-package-repositories)
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+
+- ~~Should crate owners be able to configure the allowed token scopes for a Trusted Publisher configuration?~~
+    - ~~We could default to `publish-new` and `publish-update`, but maybe it's best to allow this to be configurable?~~
+    - _Initially, all Trusted Publisher Configurations will default to `publish-update` with no option to change this. After some of the implementations in [Future Possibilities](#future-possibilities) are completed, this can be expanded to allow configuration of additional scopes._
+- ~~How long should an access token derived from the ID token exchange be valid for?~~
+    - _We will provide a long enough lifetime initially for access tokens. We may also allow this to be configured in the Trusted Publisher Configuration. Regardless of the lifetime, the action that performs the token exchange will issue a request to revoke the token at the end of the workflow.
+
+### Out of Scope
+[out-of-scope]: #out-of-scope
+
+- Provenance verification of published crates (e.g. sigstore or other signing mechanisms).
+- Discussion/debate of CI/CD providers to support next and their expected timelines.
+- Support in the GitHub Action used to perform the OIDC authentication with other Rust package registries.
+- Direct suppport and implementation of the token exchange flow within `cargo`.
+
+### Items to Resolve Before General Availability
+[resolve-before-general-availability]: #resolve-before-general-availability
+
+- Before crates.io officially supports the OIDC authentication flow from GitHub Actions, we will have an official GitHub Action to facilitate the OIDC token exchange.
+- Crates.io lacks accessible usage documentation. We will need to document the setup of a Trusted Publisher with screenshots and provide example GitHub Action workflows.
+    - Maybe a good spot for this would be in [The Cargo Book](https://doc.rust-lang.org/cargo/index.html)?
```

Comment:
FYI there is already a page for crates.io <https://doc.rust-lang.org/nightly/cargo/reference/publishing.html>, and also a page for some CI configuration examples <https://doc.rust-lang.org/nightly/cargo/guide/continuous-integration.html>.